python/ql/src/experimental/Security/CWE-090/LDAPInjection.ql

```diff
@@ -0,0 +1,87 @@
+/**
+ * @name Python LDAP Injection
+ * @description Python LDAP Injection through search filter
+ * @kind path-problem
+ * @problem.severity error
+ * @id python/ldap-injection
+ * @tags experimental	
+ *       security	
+ *       external/cwe/cwe-090
+ */
+
+// Determine precision above
+import python
+import semmle.python.dataflow.new.RemoteFlowSources
+import semmle.python.dataflow.new.DataFlow
+import semmle.python.dataflow.new.TaintTracking
+import semmle.python.dataflow.new.internal.TaintTrackingPublic
+import DataFlow::PathGraph
+
+class InitializeSink extends DataFlow::Node {
+  InitializeSink() {
+    exists(SsaVariable initVar, CallNode searchCall |
+      // get variable whose value equals a call to ldap.initialize
+      initVar.getDefinition().getImmediateDominator() = Value::named("ldap.initialize").getACall() and
+      // get the Call in which the previous variable is used
+      initVar.getAUse().getNode() = searchCall.getNode().getFunc().(Attribute).getObject() and
+      // restrict that call's attribute (something.this) to match %search%
+      searchCall.getNode().getFunc().(Attribute).getName().matches("%search%") and
+      // set the third argument (search_filter) as sink
+      this.asExpr() = searchCall.getArg(2).getNode()
+      // set the first argument (DN) as sink
+      // or this.asExpr() = searchCall.getArg(0) // Should this be set?
+    )
+  }
+}
+
+class ConnectionSink extends DataFlow::Node {
+  ConnectionSink() {
+    exists(SsaVariable connVar, CallNode searchCall |
+      // get variable whose value equals a call to ldap.initialize
+      connVar.getDefinition().getImmediateDominator() = Value::named("ldap3.Connection").getACall() and
+      // get the Call in which the previous variable is used
+      connVar.getAUse().getNode() = searchCall.getNode().getFunc().(Attribute).getObject() and
+      // restrict that call's attribute (something.this) to match %search%
+      searchCall.getNode().getFunc().(Attribute).getName().matches("%search%") and
+      // set the second argument (search_filter) as sink
+      this.asExpr() = searchCall.getArg(1).getNode()
+      // set the first argument (DN) as sink
+      // or this.asExpr() = searchCall.getArg(0) // Should this be set?
+    )
+  }
+}
+
+class EscapeSanitizer extends DataFlow::Node {
+  EscapeSanitizer() {
+    exists(Call c |
+      (
+        // avoid flow through any %escape% function
+        c.getFunc().(Attribute).getName().matches("%escape%") or // something.%escape%()
+        c.getFunc().(Name).getId().matches("%escape%") // %escape%()
+      ) and
+      this.asExpr() = c
+    )
+  }
+}
```

Comment:
Pending refactor. (Pending decision -> covering the specific methods of the libs or matching `%sanitize%` like [this](https://github.com/github/codeql/blob/df29a1636555465527d17668c19c202e365cf502/javascript/ql/src/experimental/Security/CWE-090/LdapInjectionCustomizations.qll#L69))

Reply:
Not sure how the CodeQL team feels about it (I see pros and cons) but if there are cases where a CodeQL query developer _thinks_ that data is going to be sanitized by a sanitize-type function, should the `FilterOrDNSanitizationCall` predicate be extracted out so that it can be used in other security issues and not just the LDAP Injection vulnerabilities?